File: python-algorithm/leetcode/problem_144.py
"""144. Binary Tree Preorder Traversal
https://leetcode.com/problems/binary-tree-preorder-traversal/
"""
from typing import List
from common.tree_node import TreeNode


class Solution:
    def iterative_preorder_traversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        ans = []
        s = [root]  # FILO stack
        while s:
            node = s.pop()
            ans.append(node.val)
            if node.right:
                s.append(node.right)
            if node.left:
                s.append(node.left)
        return ans

    def recursive_preorder_traversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        ans = [root.val]
        ans += self.recursive_preorder_traversal(root.left)
        ans += self.recursive_preorder_traversal(root.right)
        return ans

File: java-algo/src/main/java/com/leetcode/Problem144.java
package com.leetcode;

import com.common.TreeNode;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * 144. Binary Tree Preorder Traversal
 * https://leetcode.com/problems/binary-tree-preorder-traversal/
 */
public class Problem144 {
    /**
     * Recursion
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null)
            return ans;
        ans.add(root.val);
        if (root.left != null)
            ans.addAll(preorderTraversal(root.left));
        if (root.right != null)
            ans.addAll(preorderTraversal(root.right));
        return ans;
    }

    /**
     * Iteration.
     */
    public List<Integer> preorderTraversal2(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null)
            return ans;
        Deque<TreeNode> deque = new LinkedList<>();
        deque.offer(root);
        while (!deque.isEmpty()) {
            TreeNode node = deque.pollFirst();
            ans.add(node.val);
            if (node.right != null)
                deque.offerFirst(node.right);
            if (node.left != null)
                deque.offerFirst(node.left);
        }
        return ans;
    }
}
